### ISSUE TITLE
How to organise induced relations?

### ISSUE THREAD

**Issue by HuStmpHrrr** [agda/agda-stdlib#438]
There are a few modules implement quite general induced relations, but they seem far away:

https://github.com/agda/agda-stdlib/blob/master/src/Relation/Binary/InducedPreorders.agda
https://github.com/agda/agda-stdlib/blob/master/src/Relation/Binary/Properties/Preorder.agda
https://github.com/agda/agda-stdlib/blob/master/src/Relation/Binary/Properties/Poset.agda
https://github.com/agda/agda-stdlib/blob/master/src/Relation/Binary/NonStrictToStrict.agda

There are a few from https://github.com/agda/agda-stdlib/blob/master/src/Function/Related.agda but I figured they are not quite similar to the ones listed above, so they might just be good lying there.

So I'd imagine there can be some sort of module structures so that they are easier to find, no? Also it would help when implementing other induced relations. There won't be a question on where to put those new functions.

**Comment by MatthewDaggitt** [comment]
So I would dearly like to create a new folder `Relation.Binary.Constructions` and move the following existing modules in there:
```
Relation.Binary.On
Relation.Binary.Flip
Relation.Binary.StrictToNonStrict
Relation.Binary.NonStrictToStrict
```
to those we could add:
```
Relation.Binary.Constructions.AddMin
Relation.Binary.Constructions.AddMax
Relation.Binary.Constructions.AddMinMax
Relation.Binary.Constructions.AddNeutral
Relation.Binary.Constructions.AddNeutrals
```
where `AddMinMax` and `AddNeutrals` are the files @gallais has defined in #437.

The contents of `Relation.Binary.Properties.Preorder` and `Relation.Binary.Properties.Poset` should be superseded by #307. Actually that's a great potential solution to my current naming problem in the pull request. We can leave `Relation.Binary.Constructions.Flip` as is, and add `Relation.Binary.Constructions.Inverse` for the version without flipped equality.

Thoughts on the grand plan?

**Comment by gallais** [comment]
I don't really like the `Constructions` (especially plural) and the `Add` prefix.
What about `Relation.Binary.Free.Min` (or `Bottom`? `Min` could be misconstrued for the fact that every 2 elements are given a `min`)

**Comment by WolframKahl** [comment]
Without being familiar with all the details referred to, I just hope that you are not saying `Inverse` when you mean `Converse`...

**Comment by gallais** [comment]
> Is your suggestion of `Free` a general replacement for `Construction` or just for adding elements?

I guess all of these are indeed constructions. So it probably does make sense to have
`Construction` and `Contruction.Free` for the free ones?

As for the plural, I had a look at the list of modules and I am reminded that we are super
inconsistent when it comes to it. E.g. `Algebra.Operations` vs `Data.List.Relation`.

> Composing them will end up with one of them having two constructors though, rather than just one. Not a huge problem, but inconvenient.

True. We can probably get away with having a very simple module which does nothing
but declare a nice pattern synonym for that case & gives names to the composition of
the various lemmas talking about the orders.

**Comment by gallais** [comment]
Btw, in `Construction.Free` I think I actually like `Infimum` and `Supremum` better
than `Bottom` and `Top`. Because we can then use `Extremum` instead of `TopBottom`.

**Comment by MatthewDaggitt** [comment]
Hmm do you think `Free` is a useful subcategory to have? I personally don't feel it adds much as subclassification and I'm not sure freeness will be that important feature to the average user of `Infimum`/`Supremum`? It feels more like a property that one would prove about a construction within the file.

Maybe I'm wrong though, as always I'm open to being convinced!

**Comment by JacquesCarette** [comment]
I also think that putting the `Free` in the name does not really add much. They are indeed all `Construction`s, and some have better properties than others.

**Comment by MatthewDaggitt** [comment]
Closed with 624e077
